src/python/deepgnn/graph_engine/snark/tests/e2e_test.py

```diff
@@ -1015,25 +1132,87 @@ def test_node_sampling_distributed_graph_multiple_partitions_raises_empty_types(
     s2.reset()
 
 
+@pytest.mark.parametrize("multi_partition_graph_data", ["original"], indirect=True)
 def test_edge_sampling_distributed_graph_multiple_partitions(
     multi_partition_graph_data,
 ):
     address = [f"localhost:{find_free_port()}", f"localhost:{find_free_port()}"]
-    _ = server.Server(multi_partition_graph_data, [0], hostname=address[0])
-    _ = server.Server(multi_partition_graph_data, [1], hostname=address[1])
+    s1 = server.Server(multi_partition_graph_data, [0], hostname=address[0])
+    s2 = server.Server(multi_partition_graph_data, [1], hostname=address[1])
 
     cl = client.DistributedGraph([address[0], address[1]])
     es = client.EdgeSampler(cl, [0, 1])
     s, d, t = es.sample(size=5, seed=2)
     npt.assert_array_equal(s, [0, 0, 5, 5, 5])
     npt.assert_array_equal(d, [5, 5, 9, 9, 9])
     npt.assert_array_equal(t, [1, 1, 1, 1, 1])
+    s1.reset()
+    s2.reset()
+
+
+@pytest.mark.parametrize(
+    "storage_type",
+    [client.PartitionStorageType.memory, client.PartitionStorageType.disk],
+)
+@pytest.mark.parametrize("multi_partition_graph_data", param, indirect=True)
+def test_distributed_graph_neighbors(multi_partition_graph_data, storage_type):
+    address = [f"localhost:{find_free_port()}", f"localhost:{find_free_port()}"]
+    s1 = server.Server(
+        multi_partition_graph_data, [0], hostname=address[0], storage_type=storage_type
+    )
+    s2 = server.Server(
+        multi_partition_graph_data, [1], hostname=address[1], storage_type=storage_type
+    )
+    cl = client.DistributedGraph(address)
+    node_ids, weights, edge_types, result_counts = cl.neighbors(
+        np.array([9, 0], dtype=np.int64),
+        np.array([0, 1, 2], dtype=np.int32),
+    )
+    npt.assert_equal(node_ids, np.array([0, 5], dtype=np.int64))
+    npt.assert_almost_equal(weights, np.array([0.5, 1.0], dtype=np.float32))
+    npt.assert_equal(edge_types, np.array([0, 1], dtype=np.int32))
+    npt.assert_equal(result_counts, np.array([1, 1], dtype=np.int64))
+    s1.reset()
+    s2.reset()
+
+
+@pytest.mark.parametrize(
+    "storage_type",
+    [client.PartitionStorageType.memory, client.PartitionStorageType.disk],
+)
+@pytest.mark.parametrize("multi_partition_graph_data", param, indirect=True)
+def test_distributed_graph_node_types(multi_partition_graph_data, storage_type):
+    address = [f"localhost:{find_free_port()}", f"localhost:{find_free_port()}"]
+    s1 = server.Server(
+        multi_partition_graph_data, [0], hostname=address[0], storage_type=storage_type
+    )
+    s2 = server.Server(
+        multi_partition_graph_data, [1], hostname=address[1], storage_type=storage_type
+    )
+    cl = client.DistributedGraph(address)
+    types = cl.node_types(np.array([9, 5, 0], dtype=np.int64), -1)
```

Comment:
why do you use default type -1? Types should be only positive according to `data_spec.rst`.

Reply:
A real node cannot have type == -1, so this is a good return to signify null value.